src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -150,8 +158,31 @@ public IndexService(Injector injector, Index index, @IndexSettings Settings inde
         indexFieldData.setIndexService(this);
         bitSetFilterCache.setIndexService(this);
         this.nodeEnv = nodeEnv;
+        clusterInfoService.addListener(new ShardSizeListener());
     }
 
+    /**
+     * Just records average size of shards in the cluster as a heuristic to help when allocating new shards to data paths.
+     */
+    class ShardSizeListener implements ClusterInfoService.Listener {
```

Comment:
I think this causes a memory leak since IndexService is created per index and if it's closed you need to remove this. you should add this listener in `IndicesService`